### DISCUSSION TITLE
Trying to overcome indexing when using CocoaPods (or not)

### DISCUSSION THREAD

**Discussion by swiftyfinch** [MobileNativeFoundation/discussions#81] [General]
Hello everyone! 👋 My name is Vyacheslav, Slava for short.

I would like to up the topic of **slow indexation** in a huge Xcode project. I've written [the post](https://swiftyfinch.github.io/en/2021-03-09-rugby-story/) about this and made my tool 🏈 [Rugby](https://github.com/swiftyfinch/Rugby). The idea is to cache some pods which I rarely need to modify.

And I wonder if somebody can share their **advice**, **best practices**. How you are living with that? 😀

I will be happy to know any alternatives.
Thanks for reading and all the best 👍

**Comment by tcamin** [comment]
Hi Slava!

together with slow indexing speeds we were also experiencing slow compilation times on our mid-sized mixed Objective-C/Swift project with approximately 100 modules/pods. 

I've built a custom tools on top of CocoaPods, PodBuilder (https://github.com/Subito-it/PodBuilder), which helped us to significantly improve both of the above problems. The tools does basically 2 things:
- it prebuilds pods into frameworks (or static libs if neeeded), reducing the number of files to compile and index
- simplifies switching between the prebuilt and the original pod, when debugging or modifications to the dependencies are needed.

The prebuilt frameworks can then be pushed to a repo using git LFS effectively becoming a compile cache of the project's dependencies. The nice thing of this solution is that it is completely trasparent to the user which on any commit has up-to-date and synched build artifacts ready to use by simply git pulling. Frameworks contain all archs so compilation is snappy both on simulators and physical devices.

PodBuilder uses CocoaPods to integrate prebuilt pods by generating a local podspec with the same name of the original podspec containing the prebuilt framework as `vendored_framework`. It's like using CocoaPods to prebuild CocoaPods 😃.
This allows to easily integrate the prebuilt items in the project because we still rely on CocoaPods doing the heavy work of dependency resolution. At the same time we can also use CP to do inverse dependencies resolution which allows to determine all the pods that require to be rebuilt.

Compared to other similar solution PodBuilder is highly customizable allowing for example fine grain adjustment of build settings per pod. Additionally it has some experimental features like prebuilt item debugging, which can be achieved by tweaking the lldbinit file (https://github.com/Subito-it/PodBuilder#install_sources-command). Recently I added also support for xcframeworks.

👋 Tomas

**Reply by swiftyfinch** [reply]
Thank you for sharing, Tomas!

Does this change `Podfile` and `Podfile.lock`?

**Reply by tcamin** [reply]
Kind of: PodBuilder autogenerates a second Podfile with modified entries for pods that have been prebuilt and is used to generate the app’s xcworkspace. The original Podfile is moved without modifications to a separate location and is used by PodBuilder to determine which pod/version to build.

**Reply by nicolasrostan** [reply]
We did something similar, created a plugin that can build a xcframework from a Pod target, the key difference is that we generate that xcframework during the pod deploy, and upload it to an internal repository for artifacts. Then, on pod install of the main applications, we can search if that xcframework exists on this repo, if so we override the download sources step of cocoapods and get the xcframework, and modify the Specification in runtime to use `vendored_framework` instead of sources.

We added a flag to the Podfile (:precompiled => true/false) on each dependency, if that flag is true it will try to use the xcframework, if not it'll use sources.

This allows us to quickly switch between xcframeworks and sources for debugging, right now tho, we always use sources for Release builds due to some runtime issues with xcframeworks.

For some context we have around 130 pods in our main apps, and of those at least 100 are internal (we split the app in flows, and core libraries). I think this solution is pretty good because we can be agile when switching back and forth, and also allows us to gradually migrate the pods, but it has some painful issues, mainly due to the fact that we use xcframeworks instead of fat, and we have a very huge and legacy codebase. 
For a pod to support this solution has to be able to compile with swift library evolution enabled, which is a huge pain to enable for legacy code and with iOS 10 support, and some things are just broken, like interframework swift/objc interoperability. Another big disadvantage is that for a Pod to be used as an xcframework, all subspecs of that pod must be imported (because when we compiled the pod, we used all subspecs). The last point is the main reason we haven't open sourced the plugin, we have some buisness logic there that allow us to generate multiple xcframeworks per pod (one for each app that the pod will be used, iff they don't share the same subspecs). 

The reason we decided to go for xcframework instead of fat, is of course the idea that we wouldn't have to recompile all pods with each xcode/swift release, but the plugin could be easily modified to support fat frameworks instead.

This solution also allow us to precompile third party pods, upload them to the internal repository, and be automatically downloaded if marked with :precompiled => true, a global flag can also be added to the Podfile to search precompiled xcframework for all pods (including transitive dependencies), so it's not mandatory to declare the dependency on Podfile to be precompiled.

**Reply by swiftyfinch** [reply]
@tcamin Thanks a lot for your answers. It was interesting to know how you've solved this problem.
I surely will try `PodBuilder` in my free time 👍🏻

**Reply by swiftyfinch** [reply]
@nicolasrostan Much appreciated it's the curious solution 👍🏻

Is it `CocoaPods` plugin? Something like `plugin xcframeworks` in your Podfile?

It's a great idea with `precompiled` flag for switching between xcframeworks and sources for debugging. But how you know which pod should be precompiled and packed into xcframework on ci? Is it not dependent on `CocoaPods`?

If you change `precompiled` flag to `false` for debugging your Podfile would change. So then you need to change `precompiled` flag back and run `pod install` before making a new commit. Is that right? Or maybe there are other ways?

Also, as I can see there is a little drawback. In most time we don't need to prebuild and download both architectures for simulator and iOS. It can be space for improvements.

**Reply by nicolasrostan** [reply]
> Is it CocoaPods plugin? Something like plugin xcframeworks in your Podfile?

It is a plugin yes, we use bundler and internal ruby specs repository to distribute the gem, then import it with `plugin cocoapods-ourplugin` on the Podfile. This plugin has three key components:
1. Compile a pod target into xcframework (command, pod precompiled create)
2. Publish into internal repository (command pod precompiled publish)
3. Integrate into pod install command and decide which pods can be precompiled, and for those that can, download them and install them.

So the first two commands are used on mainly CI on the release workflow of a pod, (can be used locally for debugging reasons).
The integration part (pod install hooks) is used on any app that wants to consume a dependency as precompiled.

> It's a great idea with precompiled flag for switching between xcframeworks and sources for debugging. But how you know which pod should be precompiled and packed into xcframework on ci? Is it not dependent on CocoaPods?

For context, each pod of ours has its own repository, in that repository it also has a test app for that pod, a single team is responsible for that repo. When a team decides that it's time to release a new version of the pod, they will run a build/release workflow on CI.

So how do we decide? basically we have the team responsible for that pod do it, by setting a flag on the repo that our CI looks for, if that flag is true, then the commands for pod precompiled create, and publish will be executed after the normal spec lint and publish to our internal specs repo.

For a pod to be compatible with this solution it has to ensure a couple of things, which the team responsible has to check for before enabling it.
1. Is it using resource bundles? (this is a must)
2. Can be compiled with new build system and swift library evolution enabled?
3. Check that all the subspecs are imported on Podfile, this solution doesn't allow for "partial" precompiled, meaning it's all or nothing, you either precompile all subspecs or none. (This works quite well for us as we tend to have simple pods)
4. Test it locally (use the command pod precompiled create and see what happens, should generate a valid xcframework)

Assuming all that is good, they'll set the flag to true and on the next release it will also be published as an xcframework.

So what does the plugin do when pod precompiled create is run you may ask, off the top of my head, the plugin does the following.
1. Check that a pod install has been run before (using the test app Podfile) if not throw error.
2. Determine which targets are generated by the local .podspec, this is done by parsing the local .podspec into a Specification and then searching the corresponding target/s on the Pods project.
3. If more than one xcf needs to be generated then we generate an aggregate target on the Pods project that contains all the targets that need to be compiled
4. Ensure that swift library evolution is enabled.
5. Compile that new aggregate target (or a single one) into sim .framwork and device .framework
6. generate the xcframework with the corresponding slices generated on 5.

On step 5, we can have multiple targets for the cases when a pod is used on both of our main apps (MercadoLibre, MercadoPago) and has a subspec for each of those that are not compatible if imported together, so we absolutly need two different xcframeworks, one for when the /MercadoLibre subspec is used, another for when /MercadoPago is. As I said before, this is the main reason that we can't open source the code.

> If you change precompiled flag to false for debugging your Podfile would change. So then you need to change precompiled flag back and run pod install before making a new commit. Is that right? Or maybe there are other ways?

There is another way to disable them, you can use `pod install --disable-precompiled`  (this will remove ALL xcframeworks and install sources for ALL pods on that run), it will ignore the settings on the podfile without modifying it. But why would you want to commit a change in the precompiled flag if you just changed it to debug something? 

We also generate a local .lock (Pods/Precompiled.lock) this file is not versioned but allow us to detect which pods where precompiled in the previous pod install run, if any pod has changed the state (of it being precompiled or not) it will be marked as dirty and be reinstalled.

We need this .lock because Podfile.lock and Manifest.lock only track Pod/version, if we just change the precompiled flag but not the version of a pod, and rerun a pod install, it wouldn't detect any changes.

> Also, as I can see there is a little drawback. In most time we don't need to prebuild and download both architectures for simulator and iOS. It can be space for improvements.

You are right, we don't need both slices always, but for us is worth the tradeoff of the complexity of the plugin to track each slice individually vs just downloading both. We also have an internal store in which we publish internal versions of our apps and test apps, and those builds (to publish an app to the store) can be run locally or in CI, for these cases we absolutly need the devices slices.
And there is most definitely space for improvements :D

**Reply by swiftyfinch** [reply]
🤯 Thanks for this awesome description and answers to each question! 
It's interesting and can help someone else too. Nice system approach 👍

In my tool 🏈 [Rugby](https://github.com/swiftyfinch/Rugby) I cache all pods dependencies locally. But the way is much similar that you described above. I also create an aggregated target and rebuild only changed pods by separated Podfile-like files.

It's like selfish way :D not for the whole team, but only for interested and the bravest parties. But there is one benefit — it does not include any changes into the original `Podfile` and `Podfile.lock`. It just an optional step for the team.
